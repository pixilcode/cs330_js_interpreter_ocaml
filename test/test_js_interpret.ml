open Js_interpreter

(* run a test and print the result *)
let run_test (name, js) =
  print_endline ("--- " ^ name ^ " ---");

  let json = Util.run_acorn js in
  let ast = Parser.from_json_string json in
  let result = Interpreter.interpret ast in
  match result with
  | Ok (values, _env, _heap) ->
    let values = List.map S_expr.from_value values in
    let values = List.map S_expr.to_string values in
    List.iter print_endline values;
    print_newline ()
  | Error message ->
    let message = S_expr.from_error message in
    let message = S_expr.to_string message in
    print_endline message;
    print_newline ()

let tests = [
  ("simple_number", "1");
  ("unary", "!true");
  ("arithmetic", "1 + 2 - 3 * 4 / 6");
  ("relational", "1 < 2");
  ("equality", "1 == 2");
  ("logical", "true && false || false");
  ("short_circuit_or_logic", "1 == 1 || true + false");
  ("short_circuit_and_logic", "1 == 2 && true + false");
  ("error_boolean_1", "1 == 2 || 1 + 2");
  ("error_boolean_2", "1 == 1 && 1 + 2");
  ("error_boolean_3", "1 && true");
  ("error_boolean_4", "false || 3");
  ("error_arithmetic_1", "1 + true");
  ("error_arithmetic_2", "true - 1");
  ("error_arithmetic_3", "false * true");
  ("error_arithmetic_4", "true / true");
  ("error_unary_1", "!1");
  ("error_unary_2", "-true");
  ("error_unary_3", "+false");
  ("error_conditional", "1 ? 2 : 3");
  ("error_divide_by_zero", "6 / 0");

  (* local binding tests *)
  (
    "simple_binding", "
    let x = 0;
    x + 1
    "
  );

  (
    "undefined_variable", "
    let x = 0;
    y + 1
    "
  );

  (
    "multiple_definitions", "
    let x = 1;
    let y = 2;
    x + y
    "
  );

  (
    "multiple_definitions_same_decl", "
    let x = 1, y = 2;
    x + y
    "
  );

  (* function expression tests *)
  (
    "simple_function", "
    let f = function(x) { return 1 }
    f(2)
    "
  );

  (
    "function_environment", "
    let a = 1;
    let b = a;
    let f = function (y) { return b }
    f(2)
    "
  );

  (
    "function_param_shadowing", "
    let x = 1;
    let f = function (x) { return x }
    f(2)
    "
  );

  (
    "nested_function_calls", "
    let g = function (x) { return 1 }
    let f = function (x) { return g(x) }
    f(2)
    "
  );

  (
    "call_non_function_error", "
    let f = 1;
    f(2)
    "
  );

  (
    "function_arg_error", "
    let f = 1;
    f(2 / 0)
    "
  );

  (
    "does_not_exist_error", "
    let f = function (x) { return a }
    let a = 1;
    f(2)
    "
  );

  (* mutable values *)
  (*
  (
    "simple_assignment", "
    let x = 1;
    x = 2;
    x
    "
  );
  *)

  (*
  (
    "getter_setter", "
    let x = 1;
    let get = function (_) {
      return x;
    }
    let set = function (a) {
      return (x = a);
    }
    set(2);
    get(1000000) // parameter isn't important
    "
  )
  *)
]

(* iterate over all the tests and run them *)
let () =
  List.iter run_test tests