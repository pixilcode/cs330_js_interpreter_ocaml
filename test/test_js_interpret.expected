--- simple_number ---
(value (number 1))

--- unary ---
(value (boolean false))

--- arithmetic ---
(value (number 1))

--- relational ---
(value (boolean true))

--- equality ---
(value (boolean false))

--- logical ---
(value (boolean false))

--- short_circuit_or_logic ---
(value (boolean true))

--- short_circuit_and_logic ---
(value (boolean false))

--- error_boolean_1 ---
(error "Oh bananas! Invalid logical operation")

--- error_boolean_2 ---
(error "Oh bananas! Invalid logical operation")

--- error_boolean_3 ---
(error "Oh bananas! Invalid logical operation")

--- error_boolean_4 ---
(error "Oh bananas! Invalid logical operation")

--- error_arithmetic_1 ---
(error "Oh bananas! Invalid arithmetic operation")

--- error_arithmetic_2 ---
(error "Oh bananas! Invalid arithmetic operation")

--- error_arithmetic_3 ---
(error "Oh bananas! Invalid arithmetic operation")

--- error_arithmetic_4 ---
(error "Oh bananas! Invalid arithmetic operation")

--- error_unary_1 ---
(error "Oh bananas! Invalid unary operation")

--- error_unary_2 ---
(error "Oh bananas! Invalid unary operation")

--- error_unary_3 ---
(error "Oh bananas! Invalid unary operation")

--- error_conditional ---
(error "Oh bananas! Invalid condition in conditional")

--- error_divide_by_zero ---
(error "Oh bananas! Cannot divide by zero")

--- simple_binding ---
(value (number 1))

--- undefined_variable ---
(error "Oh bananas! Undefined identifier")

--- multiple_definitions ---
(value (number 3))

--- multiple_definitions_same_decl ---
(value (number 3))

--- simple_function ---
(value (number 1))

--- function_environment ---
(value (number 1))

--- function_param_shadowing ---
(value (number 2))

--- nested_function_calls ---
(value (number 1))

--- call_non_function_error ---
(error "Oh bananas! Non-function item cannot be called")

--- function_arg_error ---
(error "Oh bananas! Cannot divide by zero")

--- does_not_exist_error ---
(error "Oh bananas! Undefined identifier")

--- simple_assignment ---
(value (void))
(value (number 2))

--- multiple_assignment ---
(value (void))
(value (void))
(value (number 3))
(value (number 4))
(value (number 7))

--- getter_setter ---
(value (void))
(value (number 2))
(value (void))
(value (number 3))

--- recursion ---
(value (void))
(value (number 3))

